Subscribers List Details: Update People List UI

staskus opened the pull request:
Fixes #23046

- Update the People List UI so it can be used when navigating from the Subscribers card.
- Navigate to WP.com/subscribers page when a subscriber is selected.
- Remove Email followers tab.

Figma designs: IqhXWz3Iir7RMb5XH5gGfZ-fi-1137_3593

## To test:

1. Enable `Subscribers and Traffic tab` feature flag
2. Open Stats -> Subscribers
3. On Subscriber list select "View more"
4. Confirm People -> Subscribers is opened with an updated UI
5. Tap on Subscriber
6. Confirm a web view with subscriber information is opened
7. Switch to People -> Users tab and confirm an updated UI appears

| 1 | 2 | 3 | 4 |
|--------------|--------------|--------------|--------------|
| ![Simulator Screenshot - iPhone 15 - 2024-04-29 at 17 51 59](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/160a6950-77c5-4ec2-a92d-02d28c50b972) | ![Simulator Screenshot - iPhone 15 - 2024-04-29 at 17 50 04](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/bbb21198-8ba1-42bb-94a9-19e12e467a42) | ![Simulator Screenshot - iPhone 15 - 2024-04-29 at 17 49 52](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/12c556a2-0fd3-42f0-bd61-d1f3b9b0d09d) | ![Simulator Screenshot - iPhone 15 - 2024-04-29 at 17 49 33](https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/6e2a5a93-2ad3-420e-a73f-d93ff403e8a6) |

## Regression Notes
1. Potential unintended areas of impact

Existing People views, especially People -> Users would continue working as before. 

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

staskus committed:
Remove emailFollowers from People

Email followers are now aggregated together with Followers

staskus committed:
Update PeopleCell UI

staskus committed:
Update PeopleCell fonts and colors

staskus committed:
Update Role badge background and text colors

staskus committed:
Remove tableView separator

staskus committed:
Allow opening PersonViewController from code

staskus committed:
Open PersonViewController from code

staskus committed:
Open Subscribers web view after subscriber is selected in PeopleViewController

staskus committed:
Update PeopleViewController.swift

staskus committed:
Show loading indicator when switching between People tabs

staskus committed:
Adjust dynamic type for PeopleCell

Enable dynamic type for username and id, and disable it for role badge

staskus committed:
Update UI tests

guarani commented on WordPress/Classes/Services/PeopleService.swift:
```diff
@@ -76,31 +76,6 @@ struct PeopleService {
         })
     }
 
-    /// Loads a page of Email Followers associated to the current blog, starting at the specified offset.
-    ///
-    /// - Parameters:
-    ///     - offset: Number of records to skip.
-    ///     - count: Number of records to retrieve. By default set to 20.
-    ///     - success: Closure to be executed on success with the number of followers retrieved and a bool indicating if more are available.
-    ///     - failure: Closure to be executed on failure.
-    ///
-    func loadEmailFollowersPage(_ offset: Int = 0,
-                                count: Int = 20,
-                                success: @escaping ((_ retrieved: Int, _ shouldLoadMore: Bool) -> Void),
-                                failure: ((Error) -> Void)? = nil) {
-        let page = (offset / count) + 1
-        remote.getEmailFollowers(siteID, page: page, max: count, success: { followers, hasMore in
-            self.coreDataStack.performAndSave({ context in
-                self.mergePeople(followers, in: context)
-            }, completion: {
-                success(followers.count, hasMore)
-            }, on: .main)
-        }, failure: { error in
-            DDLogError(String(describing: error))
-            failure?(error)
-        })
-    }
-
```
We could also remove [`getEmailFollowers`](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/ef189077aa96219eb7a278d70481bf012f5244e6/Sources/WordPressKit/Services/PeopleServiceRemote.swift#L100-L135) from WordPressKit, since this function is no longer called.

guarani commented on WordPress/Classes/ViewRelated/People/PersonViewController.swift:
```diff
@@ -298,23 +297,6 @@ private extension PersonViewController {
-            return
-        }
-
-        service?.deleteEmailFollower(emailFollower, failure: { [weak self] error in
```
nitpick: I think `deleteEmailFollower` can now be deleted from `PeopleServiceRemote` and `PeopleService` (WordpressKit).

guarani commented on WordPress/Classes/Models/Role.swift:
```diff
@@ -23,16 +23,29 @@ extension Role {
 
 extension Role {
-    @objc var color: UIColor {
+    @objc var backgroundColor: UIColor {
```
nitpick: Now that there are fewer colors, it might make sense to group the cases for each color. I'd consider removing this `backgroundColor` entirely since it's only called in one place (in the `roleBackgroundColor` computed property).

guarani commented on WordPress/Classes/Models/Role.swift:
```diff
@@ -23,16 +23,29 @@ extension Role {
+        }
+    }
+
+    @objc var textColor: UIColor {
```
nitpick: Same as `backgroundColor` (this is only used by the `roleTextColor` computed property).

guarani commented on WordPress/Classes/ViewRelated/People/WPStyleGuide+People.swift:
nitpick: We could remove this file and use the Design System directly instead of `WPStyleGuide`. Each of the properties in this file is used very few times.

guarani reviewed (approved):
Tested and works (and looks) well!

I left some nitpicks for removing code and perhaps simplifying some of the modified code.

Out of scope here, but I noticed a few things that seem off on the People screen:
- Removing a subscriber from within the web view appears to work, but they remain in the Subscriber tab (even after a pull-to-refresh).
- I'm not too clear on what's a Subscriber and what's a Viewer. I have a private site and added a Subscriber from the app. They were added as a Viewer. This kind of makes sense, but I wonder if there is a bug here since it isn't intuitive.

staskus commented:
Thanks for the review @guarani! I'll check what can and makes sense to be addressed now. 

> Removing a subscriber from within the web view appears to work, but they remain in the Subscriber tab (even after a pull-to-refresh).

I'll check what it's about but I have an idea. People view is implemented by saving and retrieving People objects from core data. I think by removing it in webview we're not indicating in any way that it should be removed from core data.

It might make sense to keep showing an old details view for now...

> I'm not too clear on what's a Subscriber and what's a Viewer. I have a private site and added a Subscriber from the app. They were added as a Viewer. This kind of makes sense, but I wonder if there is a bug here since it isn't intuitive.

I think that going into unexplored People view's territory has proven to be a shaky decision. People view deserves an analysis and a small revamp project on its own. A GitHub task or a P2 outlining current issues and possible options could be an option. It could be added into betting table although I don't feel it would have high-enough of a priority.

staskus committed:
Remove WPStyleGuide+People

staskus committed:
Avoid PeopleCell wrong avatar appearing when scrolling

staskus committed:
Remove background and textColor from role

staskus committed:
Refresh Subscribers view after closing Subscribers details to account for removed subscribers

staskus committed:
Merge branch 'trunk' into fix/23046-subscribers-list-card-details-update-ui

staskus commented:
@guarani, fixed:

- [x] Removing a subscriber from within the web view appears to work, but they remain in the Subscriber tab (even after a pull-to-refresh).
- [x] nitpick: Now that there are fewer colors, it might make sense to group the cases for each color. I'd consider removing this backgroundColor entirely since it's only called in one place (in the roleBackgroundColor computed property).
- [x] nitpick: Same as backgroundColor (this is only used by the roleTextColor computed property).
- [x] nitpick: We could remove this file and use the Design System directly instead of WPStyleGuide. Each of the properties in this file is used very few times.

Leaving to be done later/separately:
- [ ] I'm not too clear on what's a Subscriber and what's a Viewer. I have a private site and added a Subscriber from the app. They were added as a Viewer. This kind of makes sense, but I wonder if there is a bug here since it isn't intuitive.
- [ ] nitpick: I think deleteEmailFollower can now be deleted from PeopleServiceRemote and PeopleService (WordpressKit).
- [ ] We could also remove [getEmailFollowers](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/ef189077aa96219eb7a278d70481bf012f5244e6/Sources/WordPressKit/Services/PeopleServiceRemote.swift#L100-L135) from WordPressKit, since this function is no longer called.

osullivanchris commented:
Looks really good.

- Maybe a small bit of top padding (4-8pt) above the list. The first item looks a bit tight to the top tabs
- maybe white rather than grey page background? Fresher

staskus commented:
> Maybe a small bit of top padding (4-8pt) above the list. The first item looks a bit tight to the top tabs
> maybe white rather than grey page background? Fresher

Thanks for the comments, @osullivanchris. I'll make improvements. 👍

staskus committed:
Merge branch 'trunk' into fix/23046-subscribers-list-card-details-update-ui

staskus committed:
Update People table view color

staskus committed:
Set top padding for People table
